### ISSUE TITLE
Move data dir to SD crashes the app

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#7533]
- Go to Settings - System
- Select to move the geocache data to the external SD

- Crash

This was reported for the market version 2019.04.16 on Android 4.4.2 but seems also to happen on my Android 9 device

Fix for this should target releas branch.

Not sue if this is an unfixed Targetsdk regression or something else.
Need to take a deeper look next days, unless someone is faster.

**Comment by Lineflyer** [comment]
Definitely a remaining SDK26 regression. Maybe the same problem we had with the management of other file operations.
Raising priority to high.

**Comment by Lineflyer** [comment]
@pstorch
As you implemented the related code AFAIR, could I ask you to take a look if time permits and you are interested?

IMHO changing the targetSDK broke the function due to some changes in file management.

**Comment by moving-bits** [comment]
It's some layout inflating problem. LogCat is complaining:

```
   android.view.InflateException: Binary XML file line #26: Binary XML file line #44: Error inflating class <unknown>
    Caused by: android.view.InflateException: Binary XML file line #44: Error inflating class <unknown>
    Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:430)
        at android.view.LayoutInflater.createView(LayoutInflater.java:645)
[...]
        at android.app.AlertDialog.onCreate(AlertDialog.java:423)
        at android.app.ProgressDialog.onCreate(ProgressDialog.java:198)
        at android.app.Dialog.dispatchOnCreate(Dialog.java:389)
        at android.app.Dialog.show(Dialog.java:293)
        at cgeo.geocaching.activity.Progress.show(Progress.java:55)
        at cgeo.geocaching.storage.LocalStorage.changeExternalPrivateCgeoDir(LocalStorage.java:336)
[...]
   Caused by: java.lang.UnsupportedOperationException: Failed to resolve attribute at index 3: TypedValue{t=0x2/d=0x1010036 a=1}
        at android.content.res.TypedArray.getColorStateList(TypedArray.java:528)
        at android.widget.TextView.<init>(TextView.java:775)
        at android.widget.TextView.<init>(TextView.java:704)
        at android.widget.TextView.<init>(TextView.java:700)
        at com.android.internal.widget.DialogTitle.<init>(DialogTitle.java:41)
        at java.lang.reflect.Constructor.newInstance0(Native Method) 
        at java.lang.reflect.Constructor.newInstance(Constructor.java:430) 
        at android.view.LayoutInflater.createView(LayoutInflater.java:645) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:787) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:727) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:858) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:821) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:861) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:821) 
        at android.view.LayoutInflater.parseInclude(LayoutInflater.java:994) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:854) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:821) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:518) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:426) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:377) 
        at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:414) 
        at com.android.internal.app.AlertController.installContent(AlertController.java:236) 
        at android.app.AlertDialog.onCreate(AlertDialog.java:423) 
        at android.app.ProgressDialog.onCreate(ProgressDialog.java:198) 
        at android.app.Dialog.dispatchOnCreate(Dialog.java:389) 
        at android.app.Dialog.show(Dialog.java:293) 
        at cgeo.geocaching.activity.Progress.show(Progress.java:55) 
        at cgeo.geocaching.storage.LocalStorage.changeExternalPrivateCgeoDir(LocalStorage.java:336) 
        at cgeo.geocaching.settings.SettingsActivity.lambda$null$4$SettingsActivity(SettingsActivity.java:333) 
        at cgeo.geocaching.settings.-$$Lambda$SettingsActivity$xe5YyUxpWGDqicGAfhf2okVoqDw.onClick(lambda) 
        at com.android.internal.app.AlertController$ButtonHandler.handleMessage(AlertController.java:164) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:154) 
        at android.app.ActivityThread.main(ActivityThread.java:6077) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:866) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:756) 
```

I was able to partially fix it by adding specific values to `?text_color` parameters of `cgeoProgressdialogTheme` and `cgeoProgressdialogTheme_light` in `styles.xml`, but still get error messages, this time happening at

```
    android.view.InflateException: Binary XML file line #80: Binary XML file line #37: Error inflating class <unknown>
    Caused by: android.view.InflateException: Binary XML file line #37: Error inflating class <unknown>
```

Any idea on how to find out which attribute is missing?

**Comment by pstorch** [comment]
I tried to reproduce it, but this error doesn't show up on my device. I also tried to create an emulator with 4.4.2, but there it did also work. How can I help?

**Comment by Lineflyer** [comment]
@pstorch Did you try on Oreo as well?

**Comment by pstorch** [comment]
Ah now I can reproduce it. Had to fake a second dir to enable the chooser dialog.

But I think the error is buried in one of the android libraries. I tried to upgrade to 28, but jumped from error to error, from one lib incompatibility to the next. Stopping for now.

**Comment by Lineflyer** [comment]
Does anyone else have any idea about that? It was working until we last time raised the targetSDK. From my simple thinking this is only copying around some files...no idea why this is no longer working. 

Still many users contacting us due to this brokem major feature.

**Comment by moving-bits** [comment]
Last time I've dug into this I stopped at some layout inflation problem I could not resolve.

Anybody can tell me how to find out which resource is missing when I get the following error:

```
android.view.InflateException: Binary XML file line #80: Binary XML file line #37: Error inflating class <unknown>
    Caused by: android.view.InflateException: Binary XML file line #37: Error inflating class <unknown>
```

Which "binary XML file" is referenced? Can I un-binary it? Any other way to find out which resource / attribute etc. is missing?

**Comment by moving-bits** [comment]
After some more debugging...: Crash happens on the last step of switching the data directory while opening the progress dialog. See `activity/Progress.java` in `show()` method. Caught it after adding some `try { ... } catch ...` statements:

```
    public synchronized void show(final Context context, final String title, final String message, final int style, final Message cancelMessage) {
        if (!isShowing()) {
            try {
                createProgressDialog(context, title, message, cancelMessage);
            } catch (Exception e) {
                Log.e("caught exception during createProgressDialog: " + e.getMessage());
            }
            try {
                dialog.setProgressStyle(style);
            } catch (Exception e) {
                Log.e("caught exception during setProgressStyle: " + e.getMessage());
            }
            try {
                dialog.show();
            } catch (Exception e) {
                Log.e("caught exception during dialog.show: " + e.getMessage());
            }
        }
    }
```

This is being called from `changeExternalPrivateCgeoDir()` in `storage/LocalStorage.java`.

As mentioned earlier it's a layout inflating problem, so probably a missing attribute or something like this, but haven't found the missing piece yet...

**Comment by moving-bits** [comment]
BTW: On current master there is a slightly different preference structure. So instead of
> Go to Settings - System

use `Go to Settings - Offline Data`

and then

> Select to move the geocache data to the external SD
> Crash

I'm currently able to reproduce on a physical device running Android 5. Only works if an additional SD card is added (otherwise the preference entry is greyed out).

**Comment by Cappa-d** [comment]
Recently upgraded to a new mobile phone. Now this issue happens on the new one (Android Q based) and the old one I wiped and reinstalled (Android Oreo based) as I handed it down to my son.

**Comment by djelic** [comment]
Hi all, it looks like issue is caused by progress dialog styling of `android:textColorPrimary`. Removing that property solves the issue. I also noticed that is the only place where that property is used (also in light theme) so not sure if it is required here.

Since I'm not Android developer I might be talking nonsense, sorry if that's the case I just wanted to take a shot to scratch my own itch. If this helps let me know and I'll submit PR.

**Comment by Lineflyer** [comment]
> Hi all, it looks like issue is caused by progress dialog styling of android:textColorPrimary. 
That would be incredible news, indeed. I would not be surprised as such kind of styling issues we had a lot after the change which also caused this issue.
Funny enough no one got this clue... maybe somone can check

**Comment by bekuno** [comment]
>             try {
>                 createProgressDialog(context, title, message, cancelMessage);

here is
~~~
    private void createProgressDialog(final Context context, final String title, final String message, final Message cancelMessage) {
        dialog = hideAbsolute ? new CustomProgressDialog(context) : new ProgressDialog(context, Settings.isLightSkin() ? R.style.cgeoProgressdialogTheme_light : R.style.cgeoProgressdialogTheme);
~~~

And in R.style.cgeoProgressdialogTheme_light/cgeoProgressdialogTheme is android:textColorPrimary used.

**Comment by Lineflyer** [comment]
Not in `R.style.cgeoProgressdialogTheme`? Because it happens also on dark theme.

**Comment by bekuno** [comment]
> 
> 
> Not in `R.style.cgeoProgressdialogTheme`? Because it happens also on dark theme.

In both. This was stated alredy im my comment. Now it is better formatted.

See https://github.com/heinrichreimer/android-issue-reporter/issues/87#issuecomment-340909422
They configured a strings.xml for an extra android version specific folder `values-v21` .

**Comment by moving-bits** [comment]
I tried leaving out the `android:textColorPrimary` attribute for both styles, but that alone does not seem to work, c:geo still crashes. Tested on an physical Android 5 device.

**Comment by djelic** [comment]
That's odd. I tested on Android 10 AVD and Android 8 physical device and it works. Not sure how to test it on older AVDs since move data option is grayed out.

**Comment by moving-bits** [comment]
Ok, found a possible solution - had to leave out more parts. Thanks @djelic for pointing us into that direction!

A quick test with physical device running on Android 5 (API 21) seems to be ok.
I also fired up an API 19 emulator, but cannot test the dir changer there, as I don't have an extra sd card to move data to. Reloading a cache (which also uses the progress dialog, but maybe in a different way) is fine, though, on both devices.

I'll create a PR for further testing on more physical devices, if possible. => @Lineflyer?

**Comment by moving-bits** [comment]
> That's odd. I tested on Android 10 AVD and Android 8 physical device and it works.

Maybe it also depends on UI customization by the manufacturer. (The physical device I used is a Samsung device.)

> Not sure how to test it on older AVDs since move data option is grayed out.

It's only activated if there's is an extra storage (aka "SD card") to move the data to. Don't know how this could be activated for an emulated device.

**Comment by Lineflyer** [comment]
> I'll create a PR for further testing on more physical devices, if possible. => @Lineflyer?

Yep, just merge to master for test om nightlies

**Comment by moving-bits** [comment]
Closed with #8238
